File: 프로그래머스/3/389481. 봉인된 주문/봉인된 주문.js
function solution(n, bans) {

    // 26의 거듭제곱 (BigInt)
    const pow26 = Array(12).fill(0n);
    pow26[0] = 1n;
    for (let i = 1; i <= 11; i++) {
        pow26[i] = pow26[i - 1] * 26n;
    }

    // 문자열 → 순서 번호 (1-based)
    function getRank(s) {
        const L = s.length;
        let r = 0n;

        // 더 짧은 문자열 개수
        for (let l = 1; l < L; l++) {
            r += pow26[l];
        }

        // 같은 길이에서 사전순
        for (let i = 0; i < L; i++) {
            r += BigInt(s.charCodeAt(i) - 97) * pow26[L - i - 1];
        }

        return r + 1n;
    }

    // 순서 번호 → 문자열
    function getString(k) {
        let L = 1;
        while (k > pow26[L]) {
            k -= pow26[L];
            L++;
        }

        k -= 1n; // 0-based
        let res = '';

        for (let i = 0; i < L; i++) {
            const div = pow26[L - i - 1];
            res += String.fromCharCode(
                Number(k / div) + 97
            );
            k %= div;
        }

        return res;
    }

    // bans → rank 배열
    const banned = bans.map(getRank).sort((a, b) => (a < b ? -1 : 1));

    // n 보정
    let target = BigInt(n);
    for (const r of banned) {
        if (r <= target) {
            target++;
        } else {
            break;
        }
    }

    // 결과
    return getString(target);
}

File: 프로그래머스/3/389481. 봉인된 주문/봉인된 주문.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned long long ull;

ull pow26[12];

/* 문자열 → 순서 번호 (1-based) */
ull get_rank(const char* s) {
    int L = strlen(s);
    ull r = 0;

    // 더 짧은 문자열 개수
    for (int l = 1; l < L; l++) {
        r += pow26[l];
    }

    // 같은 길이에서 사전순
    for (int i = 0; i < L; i++) {
        r += (s[i] - 'a') * pow26[L - i - 1];
    }

    return r + 1;
}

/* qsort 비교 함수 */
int cmp(const void* a, const void* b) {
    ull x = *(ull*)a;
    ull y = *(ull*)b;
    if (x < y) return -1;
    if (x > y) return 1;
    return 0;
}

/* 순서 번호 → 문자열 */
char* get_string(ull k) {
    int L = 1;
    while (k > pow26[L]) {
        k -= pow26[L];
        L++;
    }

    k--;  // 0-based
    char* res = (char*)malloc(L + 1);
    res[L] = '\0';

    for (int i = 0; i < L; i++) {
        ull div = pow26[L - i - 1];
        res[i] = 'a' + (k / div);
        k %= div;
    }

    return res;
}

char* solution(long long n, const char* bans[], size_t bans_len) {
    // 26의 거듭제곱 미리 계산
    pow26[0] = 1;
    for (int i = 1; i <= 11; i++) {
        pow26[i] = pow26[i - 1] * 26;
    }

    // bans를 rank로 변환
    ull* banned = (ull*)malloc(sizeof(ull) * bans_len);
    for (size_t i = 0; i < bans_len; i++) {
        banned[i] = get_rank(bans[i]);
    }

    // 정렬
    qsort(banned, bans_len, sizeof(ull), cmp);

    // n 보정
    ull target = (ull)n;
    for (size_t i = 0; i < bans_len; i++) {
        if (banned[i] <= target) {
            target++;
        } else {
            break;
        }
    }

    free(banned);

    // 최종 문자열 반환
    return get_string(target);
}
